---
title: "HIV Oral 16S Amplicon Analysis"
author: "Scott A. Handley"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
---

# Project Description:
Analysis of 16S rRNA amplicons in oral samples (saliva) collected from HIV infected patients pre-ART treatment and 24 weeks after treatment initiation.

This document describes the analysis of dada2 resolved sequences analysis of dada2 resolved seuqnece variants of 70 samples; 35 samples were collected at the inititation of the study from male HIV positive patients (baseline) and 35 follow-up samples 24 weeks after the inititation of ART.

**Collaborator:**
Rachel Presti (prestir@wustl.edu)

**Other relevant documents:**
Preprocessing steps can be viewed in **HIV_oral_preprocessing.Rmd**.

**References:**
* http://f1000research.com/articles/5-1492/v1
* http://benjjneb.github.io/dada2/tutorial.htm 
* Read scaling and ADONIS scripts taken from [michberr]https://github.com/michberr: https://github.com/michberr/MicrobeMiseq/blob/master/R/
* CCA code adopted from [michberr]https://github.com/michberr: http://deneflab.github.io/MicrobeMiseq/demos/mothur_2_phyloseq.html


```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=8,
                      fig.height=6,
                      fig.path="./figures/",
                      warning=FALSE,
                      message=FALSE)
```
Initiate environment

```{r initiate-environment}
library("ggplot2")
packageVersion("ggplot2")
library("phyloseq")
packageVersion("phyloseq")
library("magrittr")
packageVersion("magrittr")
library("RColorBrewer")
packageVersion("RColorBrewer")
library("vegan")
packageVersion("vegan")
library("gridExtra")
packageVersion("gridExtra")
library("knitr")
packageVersion("knitr")
library("dplyr")
packageVersion("dplyr")
library("DESeq2")
packageVersion("DeSeq2")
library("plotly")
packageVersion("plotly")
library("picante")
packageVersion("picante")
library("microbiome")
packageVersion("microbiome")
library("ggpubr")
packageVersion("ggpubr")
library("randomForest")
packageVersion("randomForest")
library("HH")
packageVersion("HH")
library("data.table")
packageVersion("data.table")

```

```{r global-theme-settings, include=FALSE}
# Set global theming
theme_set(theme_bw(base_size = 10,
                   base_family = "Helvetica"))

```
Load phyloseq data generated from dada2.

```{r initiate-data}
# Load Phyloseq Object generated in HIV_oral_preprocessing.Rmd
ps0 <- readRDS("~/Dropbox/Research/AIDS/HIV/presti/HIV_oral/dada2/data/ps0.HIV_oral.rdp.RDS")
ps0

# Load mapping file
map <- import_qiime_sample_data("~/Dropbox/Research/AIDS/HIV/presti/HIV_oral/dada2/data/HIV_oral_mapping.txt")
ps0 <- merge_phyloseq(ps0, map)

# View sample variables
sample_variables(ps0)
summary(sample_sums(ps0))
sd(sample_sums(ps0))

```
Factor reordering and renaming

```{r factor-adjustments}
# Factor re-ordering, relabelling, etc.

# Remove the _ from Week 24
sample_data(ps0)$Week
sample_data(ps0)$Week <- factor(sample_data(ps0)$Week, labels = c("Baseline", "Week 24"))
sample_data(ps0)$Week

# Reorder and relabel CD4 category factors
sample_data(ps0)$cd4_category
sample_data(ps0)$cd4_category  <- factor(sample_data(ps0)$cd4_category, levels = c("low", "high"))
sample_data(ps0)$cd4_category

sample_data(ps0)$cd4_category <- factor(sample_data(ps0)$cd4_category, labels = c("CD4 Low", "CD4 High"))
sample_data(ps0)$cd4_category

# Reorder and relabel VL category factors
sample_data(ps0)$vl_category
sample_data(ps0)$vl_category  <- factor(sample_data(ps0)$vl_category, levels = c("low", "high"))
sample_data(ps0)$vl_category

sample_data(ps0)$vl_category <- factor(sample_data(ps0)$vl_category, labels = c("VL Low", "VL High"))
sample_data(ps0)$vl_category

```
Summary Statistics

```{r data-assessment}
# Create a new data frame of the sorted row sums, a column of sorted values from 1 to the total number of individuals/counts for each RSV and a categorical variable stating these are all RSVs.
readsumsdf = data.frame(nreads = sort(taxa_sums(ps0), TRUE), 
                        sorted = 1:ntaxa(ps0),
                        type = "RSVs")


# Add a column of sample sums (total number of individuals per sample)
readsumsdf = rbind(readsumsdf,
                   data.frame(nreads = sort(sample_sums(ps0), TRUE),
                              sorted = 1:nsamples(ps0),
                              type = "Samples"))

# Make a data frame with a column for the read counts of each sample for histogram production
sample_sum_df <- data.frame(sum = sample_sums(ps0))

# Make plots
# Generates a bar plot with # of reads (y-axis) for each taxa. Sorted from most to least abundant
# Generates a second bar plot with # of reads (y-axis) per sample. Sorted from most to least
p.reads = ggplot(readsumsdf, aes(x = sorted, y = nreads)) +
  geom_bar(stat = "identity") +
  ggtitle("RSV Assessment") +
  scale_y_log10() +
  facet_wrap(~type, scales = "free") +
  ylab("# of Reads")

# Histogram of the number of Samples (y-axis) at various read depths
p.reads.hist <- ggplot(sample_sum_df, aes(x = sum)) + 
  geom_histogram(color = "black", fill = "firebrick3", binwidth = 150) +
  ggtitle("Distribution of sample sequencing depth") + 
  xlab("Read counts") +
  ylab("# of Samples")

# Final plot, side-by-side
grid.arrange(p.reads, p.reads.hist, ncol = 2)

# Basic summary statistics
summary(sample_sums(ps0))

```
Bad sample identification and removal.

```{r sample-removal-identification}
# Format a data table to combine sample summary data with sample variable data
ss <- sample_sums(ps0)
sd <- as.data.frame(sample_data(ps0)) # useful to coerce the phyloseq object into an R data frame
ss.df <- merge(sd, data.frame("RSVs" = ss), by ="row.names") # merge ss with sd by row names. Rename ss to RSVs in the new data frame

# Plot the data by the treatment variable

y = 50 # Set a threshold for the minimum number of acceptable reads. Can start as a guess
x = "Week" # Set the x-axis variable you want to examine
label = "Description" # This is the label you want to overlay on the points that are below threshold y. Should be something sample specific

p.ss.violin <- ggplot(ss.df, aes_string(x, y = "RSVs", fill = x)) + # x is what you assigned it above
  geom_boxplot(outlier.colour="NA") +
  scale_y_log10(breaks = c(0,100,1000,10000)) +
  geom_hline(yintercept = y, lty = 2) + # Draws a dashed line across the threshold you set above as y
  geom_jitter(alpha = 0.5, width = 0.15) +
  #geom_text(data = subset(ss.df, RSVs <= y), aes_string(x, y="RSVs", label=label), size=2, hjust -1) +# This labels a subset that fall below threshold variable y and labels them with the label variable
  ggtitle("No. of RSVs per Sampling Time") +
  ylab("RSVs (log10)") +
  theme(legend.position = "NULL")
p.ss.violin

# Remove samples with fewer than 50 sequences
ps0 <- prune_samples(sample_sums(ps0)>=50, ps0)
ps0
min(sample_sums(ps0))

# Remove paired sample
ps0 <- subset_samples(ps0, patient_ID != 83772)
ps0

```

```{r outlier-sample-evaluation}
# Outlier evaluation
out.bray <- ordinate(ps0, method = "MDS", distance = "bray")
p.MDS.outlier <- plot_ordination(ps0, out.bray, color = "Week", axes = c(1,2)) +
  theme_bw() +
  geom_point(size = 2) +
  ggtitle("MDS of Bray Distances \nOutlier Evaluation") +
  stat_ellipse(type = "norm", geom = "polygon", alpha = 1/10, aes(fill = Week))
p.MDS.outlier

```

All samples are realtivaley tighlty clustered. No samples will be removed based on MDS of Bray-curtis distances.

```{r prevalence-assessment}
# Begin by removing sequences that were not classified as Bacteria or were classified as either mitochondria or chlorplast

ps0 # Check the number of taxa prior to removal
ps1 <- ps0 %>%
  subset_taxa(
    Kingdom == "Bacteria" &
    Family  != "mitochondria" &
    Class   != "Chloroplast"
  )
ps1 # Confirm that the taxa were removed

ps1.base <- subset_samples(ps1, Week == "Baseline")
ps1.base
ps1.24 <- subset_samples(ps1, Week == "Week 24")
ps1.24

# Note the prefix (e.g. k__, f__, c__, etc.) is a GreenGenes convention. If you are using RDP or Silva you can simply remove these prefixes.

# To remove RSVs lacking phyla classification
#ps2 <- subset_taxa(ps2, !is.na(Phylum) & !Phylum %in% c("","uncharacterized"))

# Prevelance estimation
# Calculate feature prevelance across the data set
prevdf.base <- apply(X = otu_table(ps1.base),MARGIN = ifelse(taxa_are_rows(ps1.base), yes = 1, no = 2),FUN = function(x){sum(x > 0)})
prevdf.24 <- apply(X = otu_table(ps1.24),MARGIN = ifelse(taxa_are_rows(ps1.24), yes = 1, no = 2),FUN = function(x){sum(x > 0)})

# Add taxonomy and total read counts to prevdf
prevdf.base <- data.frame(Prevalence = prevdf.base, TotalAbundance = taxa_sums(ps1.base), tax_table(ps1.base))
prevdf.24 <- data.frame(Prevalence = prevdf.24, TotalAbundance = taxa_sums(ps1.24), tax_table(ps1.24))

# Reorder based on visual inspection of relative dominance of each taxa for plotting
prevdf.base$Phylum <- factor(prevdf.base$Phylum, levels = c("Bacteroidetes", "Firmicutes", "Proteobacteria", "Fusobacteria", "Actinobacteria", "Spirochaetes", "Tenericutes","Synergistetes"))
prevdf.24$Phylum <- factor(prevdf.24$Phylum, levels = c("Bacteroidetes", "Firmicutes", "Proteobacteria", "Fusobacteria", "Actinobacteria", "Spirochaetes", "Tenericutes","Synergistetes"))

# Create a table of Phylum, their mean abundances across all samples, and the number of samples they were detected in
plyr::ddply(prevdf.base, "Phylum", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})
plyr::ddply(prevdf.24, "Phylum", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})

#Prevalence plot - Baseline
prevdf1.base <- subset(prevdf.base, Phylum %in% get_taxa_unique(ps1.base, "Phylum"))
p.prevdf1.base <- ggplot(prevdf1.base, aes(TotalAbundance, Prevalence / nsamples(ps1.base),color=Family)) +
  # geom_hline(yintercept = 0.03, alpha = 0.5, linetype = 2) +
  geom_point(size = 1.5, alpha = 0.7) +
  scale_x_log10(limits = c(1,10000), breaks = c(100,10000)) +
  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_grid(~Phylum) +
  theme(legend.position="None") +
  theme(axis.title.x = element_blank())
  #ggtitle("Baseline Samples")

#Prevalence plot - Weke 24
prevdf1.24 <- subset(prevdf.24, Phylum %in% get_taxa_unique(ps1.24, "Phylum"))
p.prevdf1.24 <- ggplot(prevdf1.24, aes(TotalAbundance, Prevalence / nsamples(ps1.24),color=Family)) +
  # geom_hline(yintercept = 0.03, alpha = 0.5, linetype = 2) +
  geom_point(size = 1.5, alpha = 0.7) +
  scale_x_log10(limits = c(1,10000), breaks = c(100,10000)) +
  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_grid(~Phylum) +
  theme(legend.position="none")
  #ggtitle("Week 24 Samples")

grid.arrange(p.prevdf1.base, p.prevdf1.24, nrow = 2)

```

```{r phyla_alpha_diversity}
ps1_Bacteroidetes <- subset_taxa(ps1, Phylum == "Bacteroidetes")
ntaxa(ps1_Bacteroidetes)
ps1_Firmicutes <- subset_taxa(ps1, Phylum == "Firmicutes")
ntaxa(ps1_Firmicutes)
ps1_Proteobacteria <- subset_taxa(ps1, Phylum == "Proteobacteria")
ntaxa(ps1_Proteobacteria)
ps1_Fusobacteria <- subset_taxa(ps1, Phylum == "Fusobacteria")
ntaxa(ps1_Fusobacteria)
ps1_Actinobacteria <- subset_taxa(ps1, Phylum == "Actinobacteria")
ntaxa(ps1_Actinobacteria)
ps1_Spirochaetes <- subset_taxa(ps1, Phylum == "Spirochaetes")
ntaxa(ps1_Spirochaetes)
ps1_Tenericutes <- subset_taxa(ps1, Phylum == "Tenericutes")
ntaxa(ps1_Tenericutes)
ps1_Synergistetes <- subset_taxa(ps1, Phylum == "Synergistetes")
ntaxa(ps1_Synergistetes)

estimate_richness(ps1_Bacteroidetes, split = F, measures = "Shannon")
estimate_richness(ps1_Firmicutes, split = F, measures = "Shannon")
estimate_richness(ps1_Proteobacteria, split = F, measures = "Shannon")
estimate_richness(ps1_Fusobacteria, split = F, measures = "Shannon")
estimate_richness(ps1_Actinobacteria, split = F, measures = "Shannon")
estimate_richness(ps1_Spirochaetes, split = F, measures = "Shannon")
estimate_richness(ps1_Tenericutes, split = F, measures = "Shannon")
estimate_richness(ps1_Synergistetes, split = F, measures = "Shannon")

```
No reason to filter low-prevelance taxa at this stage.

Data transformations.

```{r data-transform}
# Transform to Realative abundances
ps1.ra <- transform_sample_counts(ps1, function(OTU) OTU/sum(OTU))

# Transform to Proportional Abundance
ps1.prop <- transform_sample_counts(ps1, function(x) min(sample_sums(ps1)) * x/sum(x))

# Log transformation moves to a more normal distribution
ps1.log <- transform_sample_counts(ps1, function(x) log(1 + x))

# View how each function altered count data
par(mfrow=c(1,4))
plot(sort(sample_sums(ps1), TRUE), type = "o", main = "Native", ylab = "RSVs", xlab = "Samples")
plot(sort(sample_sums(ps1.log), TRUE), type = "o", main = "log Transfromed", ylab = "RSVs", xlab = "Samples")
plot(sort(sample_sums(ps1.ra), TRUE), type = "o", main = "Relative Abundance", ylab = "RSVs", xlab = "Samples")
plot(sort(sample_sums(ps1.prop), TRUE), type = "o", main = "Proportional Abundance", ylab = "RSVs", xlab = "Samples")
par(mfrow=c(1,1))

# Histograms of the non-transformed data vs. the transformed data can address the shift to normality
p.nolog <- qplot(rowSums(otu_table(ps1))) + ggtitle("Raw Counts") +
  theme_bw() +
  xlab("Row Sum") +
  ylab("# of Samples")

p.log <- qplot(log10(rowSums(otu_table(ps1)))) +
  ggtitle("log10 transformed counts") +
  theme_bw() +
  xlab("Row Sum") +
  ylab("# of Samples")

grid.arrange(p.nolog, p.log, ncol = 2)

```
Subsetting

```{r subsetting}
# Subset. Always subset after taxa/sample filtering
# Baseline
ps1.base <- subset_samples(ps1, Week == "Baseline")
ps1.base
ps1.base <- prune_samples(sample_sums(ps1.base) > 0, ps1.base)
ps1.base
ps1.base.log <- transform_sample_counts(ps1.base, function(x) log(1 + x))

# Week 24
ps1.24 <- subset_samples(ps1, Week == "Week 24")
ps1.24
ps1.24 <- prune_samples(sample_sums(ps1.24) > 0, ps1.24)
ps1.24
ps1.24.log <- transform_sample_counts(ps1.24, function(x) log(1 + x))

```
##Community composition plotting

```{r community-composition-plots}
# Community composition - Baseline and Week 24 timepoints combined

# Phyla level plots
# Melt to long format (for ggploting) 
# Prune out phyla below % in each sample

ps1_phylum <- ps1 %>%
  tax_glom(taxrank = "Phylum") %>%                     # agglomerate at phylum level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  psmelt() %>%                                         # Melt to long format
  filter(Abundance > 0.03) %>%                         # Filter out low abundance taxa
  arrange(Phylum)                                      # Sort data frame alphabetically by phylum

# Convert Sample No to a factor because R is weird sometime
ps1_phylum$sample_no <- as.factor(ps1_phylum$sample_no)

# Write Phylum table
write.table(ps1_phylum, file = "./results/phylum.txt", sep="\t")

# Plot - Phylum - baseline vs. Week 24
p.comm.bar <- ggplot(ps1_phylum, aes(x = sample_no, y = Abundance, fill = Phylum)) + 
  theme_bw() +
  geom_bar(stat = "identity") +
  facet_wrap(~Week, scales = "free_x") +
  ggtitle("Abundant Phylum (> 3%) in All Samples") +
  ylab("Relative Abundance") +
  theme(axis.text.x = element_blank()) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank()) +
  scale_fill_manual(values = c('#1b9e77','#d95f02','#7570b3','#e7298a','#66a61e','#e6ab02','#a6761d','#666666')) +
  theme(legend.position = "bottom")
p.comm.bar

ggplotly(p.comm.bar)

```

```{r add-sample-data}
# Diversity
diversity <- global(ps1)
head(diversity)

sd.1 <- as.data.frame(sample_data(ps1)) # useful to coerce the phyloseq object into an R data frame
ps1.rich <- merge(sd.1, diversity, by ="row.names") # merge ss with sd by row names. Rename ss to RSVs in the new data frame

# Add divergence measurements
ps1.rich$divergence <- divergence(ps1)

# Can also combing just richness and diversity instead of entire diversity data frame
# sample_data(ps1)$Shannon <- diversity$diversities_shannon
# sample_data(ps1)$Richness <- diversity$richness_0
# colnames(sample_data(ps1))

```

```{r metadata-plots}
# log10 transform skewed viral load data
# ps1.rich$vl <- log10(ps1.rich$vl)
p.cd4 <- ggboxplot(ps1.rich, x = "Week", y = "cd4", add = "jitter", color = "Week", outlier.shape = NA) +
  stat_compare_means(method = "wilcox.test", label = "p.format") +
  ylab("CD4 (c/ml)")

p.vl <- ggboxplot(ps1.rich, x = "Week", y = "vl", add = "jitter", color = "Week", outlier.shape = NA) +
  stat_compare_means(method = "wilcox.test", label = "p.format") +
  scale_y_log10() +
  ylab("Viral Load (log10)")

grid.arrange(p.cd4, p.vl, ncol = 2)

plot_frequencies(sample_data(ps1), "Week", "cd4_category")

plot_frequencies(sample_data(ps1), "Week", "vl_category")

```

```{r alpha-diversity-week}
p.rich.week <- ggpaired(ps1.rich, x = "Week", y = "richness_0", outlier.shape = NA, id = "patient_ID", line.size = 0.2) +
  ylab("Richness") +
  theme(axis.title.x = element_blank()) +
  stat_compare_means(paired = TRUE, method = "wilcox.test")

p.sd.week <- ggpaired(ps1.rich, x = "Week", y = "diversities_shannon", outlier.shape = NA, id = "patient_ID", line.size = 0.2) +
  ylab("Shannon Diversity") +
  theme(axis.title.x = element_blank()) +
  stat_compare_means(paired = TRUE, method = "wilcox.test")

grid.arrange(p.rich.week, p.sd.week, ncol = 2)


```

```{r alpha-diversity-cd4-vl}
# CD4 category
p.rich.CD4 <- ggboxplot(ps1.rich, x = "cd4_category", y = "richness_0", outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  facet_grid(~Week) +
  ylab("Richness") +
  theme(axis.title.x = element_blank()) +
  stat_compare_means(method = "t.test", label = "p.signif")

p.shannon.CD4 <- ggboxplot(ps1.rich, x = "cd4_category", y = "diversities_shannon", outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  facet_grid(~Week) +
  ylab("Shannon Diversity") +
  theme(axis.title.x = element_blank()) +
  stat_compare_means(method = "t.test", label = "p.signif")

p.rich.VL <- ggboxplot(ps1.rich, x = "vl_category", y = "richness_0", outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  facet_grid(~Week) +
  ylab("Richness") +
  theme(axis.title.x = element_blank()) +
  stat_compare_means(method = "t.test", label = "p.signif")

p.shannon.VL <- ggboxplot(ps1.rich, x = "vl_category", y = "diversities_shannon", outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  facet_grid(~Week) +
  ylab("Shannon Diversity") +
  theme(axis.title.x = element_blank()) +
  stat_compare_means(method = "t.test", label = "p.signif")

grid.arrange(p.rich.CD4, p.shannon.CD4, p.rich.VL, p.shannon.VL, nrow = 2, ncol = 2)

```

```{r paired-alpha-diversity}
p.paired.rich.vl <- ggpaired(ps1.rich, x = "Week", y = "richness_0", id = "patient_ID", facet.by = "vl_category", oultier.shape = NA) +
  stat_compare_means(paired = TRUE) +
  ylab("Richness")

p.paired.rich.cd4 <- ggpaired(ps1.rich, x = "Week", y = "richness_0", id = "patient_ID", facet.by = "cd4_category", oultier.shape = NA) +
  stat_compare_means(paired = TRUE) +
  geom_boxplot(outlier.colour = "NA") +
  ylab("Richness")

grid.arrange(p.paired.rich.cd4, p.paired.rich.vl, nrow = 2)

```
Alpha diversity regression.

```{r alpha-diversity-regression-CD4}
# The plot_regression command from the microbiome package requires phyloseq objects
ps1.rich.base <- subset(ps1.rich, Week == "Baseline")
ps1.rich.24 <- subset(ps1.rich, Week == "Week 24")

# Richness
p.regr.rich.base.cd4 <- plot_regression(richness_0 ~ cd4, meta(ps1.rich.base), B=500, show.lm = TRUE, spag = TRUE, shade = FALSE) +
  ggtitle("Baseline") +
  ylab("Richness") +
  xlab("CD4 (c/ml)") +
  ylim(0,200) +
  annotate("text", x = 600, y = 185, label = "p = 0.9207") +
  theme(legend.position = "none")
cor.test(sample_data(ps1.rich.base)$richness_0, sample_data(ps1.rich.base)$cd4)

p.regr.rich.24.cd4 <- plot_regression(richness_0 ~ cd4, meta(ps1.rich.24), B=500, show.lm = TRUE, spag = TRUE, shade = FALSE) +
  ggtitle("Week 24") +
  ylab("Richness") +
  xlab("CD4 (c/ml)") +
  ylim(0,200) +
  annotate("text", x = 750, y = 185, label = "p = 0.1647") +
  theme(legend.position = "none")
cor.test(sample_data(ps1.rich.24)$richness_0, sample_data(ps1.rich.24)$cd4)

# Shannon Diversity
p.regr.sd.base.cd4 <- plot_regression(diversities_shannon ~ cd4, meta(ps1.rich.base), B=500, show.lm = TRUE, spag = TRUE, shade = FALSE) +
  ggtitle("Baseline") +
  ylab("Shannon Diversity") +
  xlab("CD4 (c/ml)") +
  ylim(0,5) +
  xlim(0,800) +
  annotate("text", x = 600, y = 4.5, label = "p = 0.7006") +
  theme(legend.position = "none")
cor.test(sample_data(ps1.rich.base)$diversities_shannon, sample_data(ps1.rich.base)$cd4)

p.regr.sd.24.cd4 <- plot_regression(diversities_shannon ~ cd4, meta(ps1.rich.24), B=500, show.lm = TRUE, spag = TRUE, shade = FALSE) +
  ggtitle("Week 24") +
  ylab("Shannon Diversity") +
  xlab("CD4 (c/ml)") +
  ylim(0,5) +
  xlim(0,800) +
  annotate("text", x = 600, y = 4.5, label = "p = 0.2559") +
  theme(legend.position = "none")
cor.test(sample_data(ps1.rich.24)$diversities_shannon, sample_data(ps1.rich.24)$cd4)

grid.arrange(p.regr.rich.base.cd4, p.regr.rich.24.cd4, p.regr.sd.base.cd4, p.regr.sd.24.cd4, ncol = 2, nrow = 2)

```

```{r alpha-diversity-regression-VL}
# VL
p.regr.rich.base.vl <- plot_regression(richness_0 ~ vl, meta(ps1.rich.base), B=500, show.lm = TRUE, spag = TRUE, shade = FALSE) +
  ggtitle("Richness vs. VL (c/ml) Baseline") +
  ylab("Richness") +
  xlab("VL (c/ml)") +
  annotate("text", x = 1500000, y = 175, label = "p = 0.3482") +
  ylim(0,200)
cor.test(sample_data(ps1.rich.base)$richness_0, sample_data(ps1.rich.base)$vl)
p.regr.rich.base.vl

#p.regr.rich.24.vl <- plot_regression(Richness ~ vl, meta(ps1.rich.24), B=500, show.lm = TRUE, spag = FALSE, shade = FALSE) +
 # ggtitle("Richness vs. VL (c/ml) Week 24") +
 # ylab("Richness") +
 # xlab("VL (c/ml)") +
 # annotate("text", x = 5.5, y = 325, label = "p = 0.9703") +
 # ylim(0,350)
# cor.test(sample_data(ps1.rich.24)$Richness, sample_data(ps1.rich.24)$vl)

```
Analysis of covariance.

```{r ANCOVA-CD4}
# Interaction between age, EE status and alpha diversity
# Or use hotdog to visualize the resutls
ancova(diversities_shannon~cd4*Week, data = ps1.rich)
#wcor.test(sample_data(ps0.rich.base)$Shannon, sample_data(ps0.rich.base)$cd4)
#cor.test(sample_data(ps0.rich.24)$Shannon, sample_data(ps0.rich.24)$cd4)

ancova(richness_0~cd4*Week, data = ps1.rich)
#cor.test(sample_data(ps0.rich.base)$Richness, sample_data(ps0.rich.base)$CD4)
#cor.test(sample_data(ps0.rich.24)$Richness, sample_data(ps0.rich.24)$CD4)

```

```{r ANCOVA-VL}
# Interaction between age, EE status and alpha diversity
# Or use hotdog to visualize the resutls
ancova(diversities_shannon~vl*Week, data = ps1.rich)
#cor.test(sample_data(ps0.rich.base)$Shannon, sample_data(ps0.rich.base)$VL)
#cor.test(sample_data(ps0.rich.24)$Shannon, sample_data(ps0.rich.24)$VL)

ancova(richness_0~vl*Week, data = ps1.rich)
#cor.test(sample_data(ps0.rich.base)$Richness, sample_data(ps0.rich.base)$VL)
#cor.test(sample_data(ps0.rich.24)$Richness, sample_data(ps0.rich.24)$VL)

```

```{r intra-inter}
# Intra = within (same patient)
# Inter = between (across patients)

# Convert to otu table
veganotu = function(physeq) {

  OTU = otu_table(physeq)
  if (taxa_are_rows(OTU)) {
    OTU = t(OTU)
  }
  return(as(OTU, "matrix"))
}

otu.table <- veganotu(ps1)
otu.table.base <- veganotu(ps1.base)
otu.table.24 <- veganotu(ps1.24)

dat.hell <- decostand(otu.table, method = "hellinger")
dat.hell.base <- decostand(otu.table.base, method = "hellinger")
dat.hell.24 <- decostand(otu.table.24, method = "hellinger")

dist.hell <- vegdist(dat.hell, method = "bray")
dist.hell.base <- vegdist(dat.hell.base, method = "bray")
dist.hell.24 <- vegdist(dat.hell.24, method = "bray")
write.table(as.matrix(dist.hell), file = "./results/dist_hell_bray.txt", sep = "\t")

pair.hell <- data.frame(t(combn(rownames(as.matrix(dist.hell)),2, simplify = TRUE)), as.numeric(dist.hell))
head(pair.hell)
write.table(pair.hell, file = "./results/pair_hell.txt", sep = "\t")

pair.hell.base <- data.frame(t(combn(rownames(as.matrix(dist.hell.base)),2, simplify = TRUE)), as.numeric(dist.hell.base))
head(pair.hell.base)
write.table(pair.hell.base, file = "./results/pair_hell_base.txt", sep = "\t")

pair.hell.24 <- data.frame(t(combn(rownames(as.matrix(dist.hell.24)),2, simplify = TRUE)), as.numeric(dist.hell.24))
head(pair.hell.24)
write.table(pair.hell.24, file = "./results/pair_hell_24.txt", sep = "\t")

```

Random forest feature selection.

```{r random-forest}
library(reshape2)
set.seed(2)
# Baseline samples
output.forest.base <- randomForest(cd4_category ~  diversities_shannon + richness_0 + vl + tnf1 + tnf2 + sCD14, na.action = na.omit, data = ps1.rich.base)
print(output.forest.base)

importance(output.forest.base, type = 2)
importance.molten.base <- melt( importance(output.forest.base, type = 2))
importance.molten.base.sort <- arrange(importance.molten.base, desc(value))
importance.molten.base.sort$Var1 <- factor(importance.molten.base.sort$Var1, levels = importance.molten.base.sort$Var1)

p.rf.base <- ggplot(importance.molten.base.sort, aes(x = Var1, y = value)) +
  geom_bar(stat = "identity") +
  ylab("Mean Decrease Gini (Importance)") +
  theme(axis.title.x = element_blank()) +
  scale_x_discrete(labels=c("TNF-2", "Viral Load", "Shannon Diversity", "TNF-1", "sCD14", "Richness")) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  annotate("text", x = 5, y = 2.5, label = "OOB error = 13.33%")

# Week 24 samples
output.forest.24 <- randomForest(cd4_category ~  diversities_shannon + richness_0 + vl + tnf1 + tnf2 + sCD14, na.action = na.omit, data = ps1.rich.24)
print(output.forest.24)

importance(output.forest.24, type = 2)
importance.molten.24 <- melt( importance(output.forest.24, type = 2))
importance.molten.24.sort <- arrange(importance.molten.24, desc(value))
importance.molten.24.sort$Var1 <- factor(importance.molten.24.sort$Var1, levels = importance.molten.24.sort$Var1)

p.rf.24 <- ggplot(importance.molten.24.sort, aes(x = Var1, y = value)) +
  geom_bar(stat = "identity") +
  ylab("Mean Decrease Gini (Importance)") +
  theme(axis.title.x = element_blank()) +
  scale_x_discrete(labels=c("Richness", "Shannon Diversity", "sCD14", "TNF-1", "TNF-2", "Viral Load")) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  annotate("text", x = 5, y = 2.5, label = "OOB error = 6.45%")

grid.arrange(p.rf.base, p.rf.24, ncol = 2)


```

Beta Diversity Analysis

```{r beta-diversity}
# Remove Samples with uncollected (Not Collected) viral loads
# sample_data(ps1)$vl_category
# ps0.rem_nc <- subset_samples(ps0, VL_category != "Not collected")
# sample_data(ps0.rem_nc)$VL_category

ord.ps1.wuni <- ordinate(ps1.log, method = "PCoA", distance = "wunifrac")

#  Single PCoA lot for presentation / publication
p.ord.CD4 <- plot_ordination(ps1.log, ord.ps1.wuni, color = "cd4_category") +
  geom_point(size=3, alpha = 0.7) +
  scale_color_brewer(palette = "Dark2") +
  geom_point(colour = "grey50", size = 1.5) +
  facet_grid(~ Week) +
  stat_ellipse(geom = "polygon", alpha = 1/15, aes(fill = cd4_category)) +
  labs(color = "CD4 Category")
p.ord.CD4

p.ord.week <- plot_ordination(ps1.log, ord.ps1.wuni, color = "Week") +
  geom_point(size=3, alpha = 0.7) +
  scale_color_brewer(palette = "Dark2") +
  geom_point(colour = "grey50", size = 1.5) +
  stat_ellipse(geom = "polygon", alpha = 1/15, aes(fill = Week)) +
  geom_line(aes(group = patient_ID), color = "black", lty = 2, lwd = 0.2)
p.ord.week

# p.ord.VL <- plot_ordination(ps1, ord.ps1.wuni, color = "Week") +
  # geom_point(size=3, alpha = 0.7) +
  # scale_color_brewer(palette = "Dark2") +
  # geom_point(colour = "grey50", size = 1.5) +
  # facet_grid(~ vl_category)

# grid.arrange(p.ord.CD4, p.ord.VL, nrow = 2)

 ######### geom_line(aes(group = patient_id), color = "black") +
p.connect <- plot_ordination(ps1.log, ord.ps1.wuni, color = "Week") +
  geom_point(size=3, alpha = 0.7) +
  scale_color_brewer(palette = "Dark2") +
  geom_point(colour = "grey50", size = 1.5) +
  geom_line(aes(group = patient_ID), color = "black")
p.connect

```

Premanova significance testing

```{r ADONIS}
# Set a random seed so that exact results can be reproduced
set.seed(1000)

# Function to run adonis test on a physeq object and a variable from metadata 
doadonis <- function(physeq, category) {
  bdist <- phyloseq::distance(physeq, "bray")
  col <- as(sample_data(physeq), "data.frame")[ ,category]
  
  # Adonis test
  adonis.bdist <- adonis(bdist ~ col)
  print("Adonis results:")
  print(adonis.bdist)
  
  # Homogeneity of dispersion test
  betatax = betadisper(bdist,col)
  p = permutest(betatax)
  print("Betadisper results:")
  print(p$tab)
}

# Runs Permanov and Homogeneity of dispersion test
# See ?betadisper for more info
doadonis(ps1, "Week")
?doadonis(ps1, "cd4_category")
doadonis(ps1.base, "cd4_category") 
doadonis(ps1.base, "vl_category")
doadonis(ps1.base, "prace")
doadonis(ps1.base, "sex")
doadonis(ps1.base, "pethnc")


doadonis(ps1.24, "cd4_category")
doadonis(ps1.24, "prace")
doadonis(ps1.24, "sex")
doadonis(ps1.24, "pethnc")

```
Constrained Correspondance Analysis (CCA).

```{r CCA}
# Initially, all samples were included in the CCA analysis. However, these results clearly show that in the context of CD4 counts and viral load there are two outliers that may skew interpretation. These are the baseline sample from patient 122214 and the week 24 sample from patient 83771. Patient 122214's baseline sample had an extremely low number of CD4 T-cells and a extremely high viral load. Patient 83771's week 24 sample had a very high viral load.
# Due to these factors, these samples were removed from the CCA analysis.

#ps1
#ps1_cca_out_rm <- ps0 %>%
#  subset_samples(
#    SampleID != "16750" &
#    SampleID != "16767"
#  )
#ps0_cca_out_rm

# Constrained Ordination - Baseline

# Remove data points with missing metadata
#ps0_not_na <- ps0_cca_out_rm%>%
#  subset_samples(
#    !is.na(CD4) &
#    !is.na(VL)
#      )

# Subset and log transforme not_na samples
#ps0_not_na.log <- transform_sample_counts(ps0_not_na, function(x) log(1 + x))
#ps0.not_na.base <- subset_samples(ps0_not_na.log, Week == "Baseline")
#ps0.not_na.base
#ps0.not_na.24 <- subset_samples(ps0_not_na.log, Week == "Week 24")
#ps0.not_na.24

# Calculate distance
bray.base <- phyloseq::distance(physeq = ps1.base.log, method = "bray")
bray.24 <- phyloseq::distance(physeq = ps1.24.log, method = "bray")

# CAP Ordinate
cap_ord.base <- ordinate(ps1.base.log, method = "CAP", distance = bray.base, formula = ~vl + cd4)
cap_ord.24 <- ordinate(ps1.24.log, method = "CAP", distance = bray.24, formula = ~vl + cd4)

# CAP plot - Baseline
cap_plot.base <- plot_ordination(ps1.base.log, ordination = cap_ord.base) +
  theme_bw() +
  geom_point(aes(colour = cd4_category), alpha = 0.7, size = 4) +
  geom_point(colour = "grey50", size = 1.5) +
  scale_color_brewer(palette = "Dark2") +
  ggtitle("CCA ~ CD4 T Cells + Viral Load \n Baseline") +
  theme(plot.title = element_text(hjust = 0.5)) +
  labs(colour = "CD4 Category")
  #geom_text(aes(label = patient_id), size = 3, check_overlap = TRUE, vjust = 2)

# CAP plot - Week 24
cap_plot.24 <- plot_ordination(ps1.24.log, ordination = cap_ord.24) +
  theme_bw() +
  geom_point(aes(colour = cd4_category), alpha = 0.7, size = 4) +
  geom_point(colour = "grey50", size = 1.5) +
  scale_color_brewer(palette = "Dark2") +
  ggtitle("CCA ~ CD4 T Cells + Viral Load \n Week 24") +
  theme(plot.title = element_text(hjust = 0.5)) +
  labs(colour = "CD4 Category")
  # geom_text(aes(label = patient_id), size = 3, check_overlap = TRUE, vjust = 2)

# Now add the environmental variables as arrows
arrowmat.base <- vegan::scores(cap_ord.base, display = "bp")
arrowmat.24 <- vegan::scores(cap_ord.24, display = "bp")

# Add labels, make a data.frame
arrowdf.base <- data.frame(labels = rownames(arrowmat.base), arrowmat.base)
arrowdf.24 <- data.frame(labels = rownames(arrowmat.24), arrowmat.24)

# Define the arrow aesthetic mapping
arrow_map <- aes(xend = CAP1, 
    yend = CAP2, 
    x = 0, 
    y = 0, 
    shape = NULL, 
    color = NULL, 
    label = labels)

label_map <- aes(x = 1.3 * CAP1, 
    y = 1.3 * CAP2, 
    shape = NULL, 
    color = NULL, 
    label = labels)

arrowhead = arrow(length = unit(0.02, "npc"))

# Make a new graphic
# CAP plot with arrows - baseline
cap_plot.base_arrows <- cap_plot.base + 
  geom_segment(
    mapping = arrow_map, 
    size = .5, 
    data = arrowdf.base, 
    color = "gray", 
    arrow = arrowhead
  ) + 
  geom_text(
    mapping = label_map, 
    size = 4,  
    data = arrowdf.base, 
    show.legend = FALSE
  )

# CAP plot with arrows - Week 24
cap_plot.24_arrows <- cap_plot.24 + 
  geom_segment(
    mapping = arrow_map, 
    size = .5, 
    data = arrowdf.24, 
    color = "gray", 
    arrow = arrowhead
  ) + 
  geom_text(
    mapping = label_map, 
    size = 4,  
    data = arrowdf.24, 
    show.legend = FALSE
  )

grid.arrange(cap_plot.base_arrows, cap_plot.24_arrows, ncol = 2)

anova(cap_ord.base, permutations = 9999)
anova(cap_ord.24, permutations = 9999) 

```
Differential Abundance Testing - Week.

```{r diffab-baseline-vs-week-24}
# Differential Abundance - Week
ds.week <- phyloseq_to_deseq2(ps1, ~Week)
dds.week <- DESeq(ds.week, test="Wald", fitType="parametric", betaPrior = FALSE) # Worth trying parametric and local fitTypes
res.dds.week = results(dds.week, cooksCutoff = FALSE)
alpha = 0.1 # Note: this is higher to try and allow even weak effect size taxa through
sigtab_dds.week = res.dds.week[which(res.dds.week$padj < alpha), ]
head(sigtab_dds.week) # Empty table

write.table(sigtab_dds.week, file="./results/deseq_results_week.txt", sep = "\t") # Change filename to save results to appropriate file

# Quick check of factor levels
# Factor on the right will be on the left in the volcano plots. Relevel to switch
mcols(res.dds.week, use.names = TRUE)


# Create volcano plots of signficant results
# Prepare to join results data.table and taxonomy table
# Baseline
resdt.week = data.table(as(results(dds.week, cooksCutoff = FALSE), "data.frame"),
                   keep.rownames = TRUE)
setnames(resdt.week, "rn", "OTU")
taxdt.week = data.table(data.frame(as(tax_table(ps1), "matrix")), keep.rownames = TRUE)
setnames(taxdt.week, "rn", "OTU")

# Join results data.table and taxonomy table
setkeyv(taxdt.week, "OTU")
setkeyv(resdt.week, "OTU")
resdt.week <- taxdt.week[resdt.week]
resdt.week
resdt.week[, Significant := padj < alpha]
resdt.week[!is.na(Significant)]
resdt.week

volcano.week = ggplot(
  data = resdt.week[!is.na(Significant)][(pvalue < 1)],
  mapping = aes(x = log2FoldChange,
                y = -log10(pvalue),
                color = Phylum,
                label = OTU, label1 = Genus)) +
  theme_bw() +
  geom_point() + 
  geom_point(data = resdt.week[(Significant)], size = 7, alpha = 0.7) + 
  # geom_text(data = resdt.base[(Significant)], mapping = aes(label = paste("Genus:", Genus)), color = "black", size = 3) +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5)) +
  geom_hline(yintercept = -log10(alpha)) +
  ggtitle("DESeq2 Negative Binomial Test Volcano Plot\nBaseline Samples") +
  theme(axis.title = element_text(size=12)) +
  theme(axis.text = element_text(size=12)) +
  theme(legend.text = element_text(size=12)) +
  geom_vline(xintercept = 0, lty = 2)
volcano.week

ggplotly(volcano.week)

```
#Differential Abundance Testing - Baseline Samples

```{r diffab-baseline-samples}
# Differential Abundance - Baseline ~ CD4 and VL Category
ds.base <- phyloseq_to_deseq2(ps1.base, ~cd4_category) # Change formula to VL_category and rerun to test VL_category
dds.base <- DESeq(ds.base, test="Wald", fitType="local", betaPrior = FALSE) # Worth trying parametric and local fitTypes

# Tabulate and write results
alpha = 0.1
res.dds.base = results(dds.base, cooksCutoff = FALSE)
sigtab_dds.base = res.dds.base[which(res.dds.base$padj < alpha), ]
sigtab_dds.base = cbind(as(sigtab_dds.base, "data.frame"), as(tax_table(ps1.base)[rownames(sigtab_dds.base), ], "matrix"))
head(sigtab_dds.base)

write.table(sigtab_dds.base, file="./results/deseq_results_base.txt", sep = "\t") # Change filename to save results to appropriate file

# Quick check of factor levels
# Factor on the right will be on the left in the volcano plots. Relevel to switch
mcols(res.dds.base, use.names = TRUE)

# Create volcano plots of signficant results
# Prepare to join results data.table and taxonomy table
# Baseline
resdt.base = data.table(as(results(dds.base, cooksCutoff = FALSE), "data.frame"),
                   keep.rownames = TRUE)
setnames(resdt.base, "rn", "OTU")
taxdt.base = data.table(data.frame(as(tax_table(ps1.base), "matrix")), keep.rownames = TRUE)
setnames(taxdt.base, "rn", "OTU")

# Join results data.table and taxonomy table
setkeyv(taxdt.base, "OTU")
setkeyv(resdt.base, "OTU")
resdt.base <- taxdt.base[resdt.base]
resdt.base
resdt.base[, Significant := padj < alpha]
resdt.base[!is.na(Significant)]
resdt.base

volcano.base = ggplot(
  data = resdt.base[!is.na(Significant)][(pvalue < 1)],
  mapping = aes(x = log2FoldChange,
                y = -log10(pvalue),
                color = Phylum,
                label = OTU, label1 = Genus)) +
  theme_bw() +
  geom_point() + 
  geom_point(data = resdt.base[(Significant)], size = 7, alpha = 0.7) + 
  # geom_text(data = resdt.base[(Significant)], mapping = aes(label = paste("Genus:", Genus)), color = "black", size = 3) +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5)) +
  geom_hline(yintercept = -log10(alpha)) +
  ggtitle("DESeq2 Negative Binomial Test Volcano Plot\nBaseline Samples") +
  theme(axis.title = element_text(size=12)) +
  theme(axis.text = element_text(size=12)) +
  theme(legend.text = element_text(size=12)) +
  geom_vline(xintercept = 0, lty = 2)
volcano.base

ggplotly(volcano.base)

```
#Differential Abundance Testing - Week 24 Samples

```{r diffab-week-24-samples}
# Differential Abundance - 24line ~ CD4 and VL Category
ds.24 <- phyloseq_to_deseq2(ps1.24, ~cd4_category) # Change formula to VL_category and rerun to test VL_category
dds.24 <- DESeq(ds.24, test="Wald", fitType="local", betaPrior = FALSE) # Worth trying parametric and local fitTypes

res.dds.24 = results(dds.24, cooksCutoff = FALSE)
alpha = 0.1
sigtab_dds.24 = res.dds.24[which(res.dds.24$padj < alpha), ]
head(sigtab_dds.24)

# There are no taxa differentially abundant at p = 0.1 between CD4 or VL categories at Week 24

```



Taxon specific barplots from DA taxa identified with DeSeq2.

```{r ground-truth}
replace_counts = function(physeq, dds) {

  dds_counts = counts(dds, normalized = TRUE)
  if (!identical(taxa_names(physeq), rownames(dds_counts))) {
    stop("OTU ids don't match")
  }
  otu_table(physeq) = otu_table(dds_counts, taxa_are_rows = TRUE)
  return(physeq)

}

# Plots of DeSeq2 identified taxa

# Subset DeSeq2 identified taxa
# Can choose between 2 transformations from above
# Typically go with relative abundances as these are more familiar to readers

# ps1.base.ra <- transform_sample_counts(ps1.base, function(OTU) OTU/sum(OTU))
# ps1.24.ra <- transform_sample_counts(ps1.24, function(OTU) OTU/sum(OTU))
ps1.rlog <- replace_counts(ps1, dds.week)
ps1.base.rlog <- replace_counts(ps1.base, dds.base)
ps1.24.rlog <- replace_counts(ps1.24, dds.24)

# Genus = Treponema
ps1_Treponema <- ps1.rlog %>%
    subset_taxa(Genus == "Treponema") %>%
    psmelt() 

# Genus = Muribaculum
ps1_Muribaculum <- ps1.rlog %>%
  subset_taxa(Family == "Porphyromonadaceae") %>%
  psmelt()

# Genus = Haemophilus
ps1_Haemophilus <- ps1.rlog %>%
  subset_taxa(Genus == "Haemophilus") %>%
  psmelt()

# Treponema boxplots
p.treponema <- ggboxplot(ps1_Treponema, x = "Week", y = "Abundance", outlier.shape = NA) +
  geom_jitter(aes(color = Species)) +
  scale_y_log10(breaks=c(0,1,10,100)) +
  facet_grid(~Species) +
  ylab("Abundance (rlog)") +
  theme(axis.title.x = element_blank()) +
  theme(legend.position = "NULL") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  stat_compare_means(method = "wilcox.test", label = "p.signif")
p.treponema

# Muribaculum boxplots
p.muribaculum <- ggboxplot(ps1_Muribaculum, x = "Week", y = "Abundance", outlier.shape = NA) +
  geom_jitter(aes(color = Species)) +
  scale_y_log10(breaks=c(0,1,10,100)) +
  facet_grid(~Species) +
  ylab("Abundance (rlog)") +
  theme(axis.title.x = element_blank()) +
  theme(legend.position = "NULL") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  stat_compare_means(method = "wilcox.test", label = "p.signif")
p.muribaculum

# Haemophilus boxplots
p.haemophilus <- ggboxplot(ps1_Haemophilus, x = "cd4_category", y = "Abundance", outlier.shape = NA) +
  geom_jitter(aes(color = Species)) +
  scale_y_log10(breaks=c(0,1,10,100,1000)) +
  facet_grid(Week~Species) +
  ylab("Abundance (rlog)") +
  theme(axis.title.x = element_blank()) +
  theme(legend.position = "NULL") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  stat_compare_means(method = "wilcox.test", label = "p.signif")
p.haemophilus

```

